cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,331 @@
+                   StandardCopyOption.REPLACE_EXISTING);
+      } finally {
+        tempFile.delete();
+      }
+
+      return newLocation;
+    } finally {
+      urlConn.disconnect();
+    }
+  }
+
+  /**
+   * This checks the local cache for this artifact and retrieves the timestamp for the newest cache entry, if this
+   * artifact is not cached it returns null
+   */
+  @Nullable
+  private Long getCurrentLastModifiedTimestamp(ArtifactId artifactId) {
+    File artifactDir = getArtifactDirLocation(artifactId);
+
+    // Check if we have cached jars in the artifact directory, if so return the latest modified timestamp.
+    return DirUtils.listFiles(artifactDir, File::isFile).stream()
+      .map(File::getName)
+      .map(FilenameUtils::removeExtension)
+      .map(Long::valueOf)
+      .max(Long::compare)
+      .orElse(null);
+  }
+
+  /**
+   * Helper method for handling the deletion of the old cache files, if required. This will also delete the directory
+   * containing the unpacked contents of the out-of-date jar.
+   */
+  private void deleteOldCache(ArtifactId artifactId, Long lastModifiedTimestamp, File newJarLocation) {
+    // This means we already have a jar but its out of date, we should delete the jar and the unpacked directory
+    File oldJarLocation = getArtifactJarLocation(artifactId, lastModifiedTimestamp);
+    LOG.debug("Got new location as {} with old location as {}", newJarLocation, oldJarLocation);
+    if (!newJarLocation.equals(oldJarLocation)) {
+      File oldUnpackLocation = getUnpackLocalPath(artifactId, lastModifiedTimestamp);
+      LOG.debug("Deleting previously cached jar");
+      try {
+        oldJarLocation.delete();
+        FileUtils.deleteDirectory(oldUnpackLocation);
+      } catch (IOException e) {
+        //Catch and log the exception, this should not cause the operation to fail
+        LOG.warn("Failed to delete old cached jar for artifact {} version {}: {}", artifactId.getArtifact(),
+                 artifactId.getVersion(), e);
+      }
+    }
+  }
+
+  /**
+   * Helper function for verifying, extracting and converting the Last-Modified header from the URL connection.
+   */
+  private ZonedDateTime getLastModifiedHeader(HttpURLConnection urlConn) {
+    Map<String, List<String>> headers = urlConn.getHeaderFields();
+    List<String> lastModifiedHeader = headers.entrySet().stream()
+      .filter(headerEntry -> LAST_MODIFIED_HEADER.equalsIgnoreCase(headerEntry.getKey()))
+      .map(Map.Entry::getValue)
+      .findFirst()
+      .orElse(null);
+
+    if (lastModifiedHeader == null || lastModifiedHeader.size() != 1) {
+      // This should never happen since this endpoint should always set the header.
+      // If it does happen we should retry.
+      throw new RetryableException(String.format("The response from %s did not contain the %s header.",
+                                                 urlConn.getURL(), LAST_MODIFIED_HEADER));
+    }
+
+    return LocalDateTime
```

Comment:
You can just the `ZonedDateTime` to parse directly

Reply:
done